okfn-brasil/serenata-de-amor
# Adding Prometheus and Grafana for load analysis and reimbursements analytics

**What is the purpose of this Pull Request?**


Provide a tool that is capable to monitor the performance of all the applications that are part of the serenata solution and show those informations with graphics that we can use to compare the performance of different solutions.

At the same time, with prometheus+grafana, we can provide dashboards that can be used to study the reimbursements data. Prometheus is great for analytics with their [PromQL](https://prometheus.io/docs/prometheus/latest/querying/basics/) while grafana is great for show data with a friendly interface.



**What was done to achieve this purpose?**


I've added prometheus and grafana to the docker-compose, along with some agents that are responsible to provide monitoring metrics from the applications to the prometheus server.

One agent in particular is responsible for quering postgres once a day to provide the reimbursements data to prometheus, so we can make our analytical study on that data.

**How to test if it really works?**


First start all applications
```
docker-compose up -d
```
Your prometheus will be available at http://localhost:9090
Your grafana will be available at http://localhost:3000
You can log in into grafana with the credentials *admin/admin* (It can be changed)

You can see the dashboards at http://localhost:3000/dashboards
You will find 2 folders, `Monitoring` will have all the dashboards responsible to monitor and provide load analysis of serenata's applications and `Business Inteligence` will have the dashboards the we will use to make analytics studies about the reimbursements.
PS: Sometimes the folders and dashboards takes a long time to be provisioned after you start the grafana server. If it's not there, wait 5 to 10 minutes for them to load.

When prometheus and grafana are up, you can monitor the performance of your applications at:
* Linux host -> http://localhost:3000/d/m2nXDoemz/linux?orgId=1
* Postgres -> http://localhost:3000/d/sadfsdfsfd/postgres?orgId=1
* All processes -> http://localhost:3000/d/lOXQMRMZ1/process-monitoring?orgId=1

For load analysis, for example, open the Process monitoring dashboard and then execute:
```
docker run --rm -v /tmp/serenata-data:/tmp/serenata-data serenata/rosie python rosie.py run chamber_of_deputies 
```
You will be able to see how much hardware resources each process is consuming, as show at the image below:

![image](https://user-images.githubusercontent.com/24193764/58755015-97a8a300-84b1-11e9-82d0-24511ac70743.png)

If you make changes at rosie's code, you will be able to evaluate if the performance got better or worse.

---
To validate the reimbursements stuff is a lit bit trickier, since it can take 1 whole day to prometheus to gather the data, not because it's slow, just because it gather that information only once a day.

But first you gotta populate the the database.

```
docker-compose run --rm django python manage.py migrate
docker-compose run --rm django python manage.py reimbursements /mnt/data/reimbursements_sample.csv
docker-compose run --rm django python manage.py companies /mnt/data/companies_sample.xz
```


After that you can see if prometheus got that data from postgres if you go to http://localhost:9090/targets and you see the businessInteligence job with the `UP` status. As shown in the image below:
![image](https://user-images.githubusercontent.com/24193764/58755356-d68e2700-84b8-11e9-964a-ccdf8fc129ee.png)

If you see the `UNKNOWN` status, it means that prometheus still haven't queried postgres yet. 

Once prometheus have the reimbursement data, you will be able to manipulate that information with PromQL and build dashboards like the one I made, accesible at http://localhost:3000/d/ecOrj5MZz/reimbursements-analytics?orgId=1

**Who can help reviewing it?**


Anyone capable of starting the containers.


**TODO**
